files/lib/page/ConversationListPage.class.php

```diff
@@ -149,9 +149,52 @@ protected function initObjectList() {
 		$this->objectList->setLabelList($this->labelList);
 		
 		if (!empty($this->participants)) {
-			$this->objectList->getConditionBuilder()->add('conversation.conversationID IN (SELECT conversationID FROM wcf'.WCF_N.'_conversation_to_user WHERE username IN (?) GROUP BY conversationID HAVING COUNT(conversationID) = ?)', [
+			// The condition is split into two branches in order to account for invisible participants.
+			// Invisible participants are only visible to the conversation starter and remain invisible
+			// until the write their first message.
+			//
+			// We need to protect these users from being exposed as participants by including them for
+			// any conversation that the current user has started. For all other conversations, users
+			// flagged with `isInvisible = 0` must be excluded.
+			//
+			// See https://github.com/WoltLab/com.woltlab.wcf.conversation/issues/131
+			$this->objectList->getConditionBuilder()->add('
+				(
+					(
+						conversation.userID = ?
+						AND
+						conversation.conversationID IN (
+							SELECT          conversationID
+							FROM            wcf'.WCF_N.'_conversation_to_user
+							WHERE           username IN (?)
```

Comment:
`username` is not indexed. I suggest to first fetch the `userID`s belonging to the given participant usernames. I'm aware that this is unchanged from the previous version, but now that filtering needs to be performed twice.

Do you possibly have the result of an `EXPLAIN` for the resulting query?

Reply:
```sql
EXPLAIN SELECT COUNT(*) AS COUNT FROM wcf2_conversation_to_user
conversation_to_user LEFT JOIN wcf2_conversation conversation ON
(conversation.conversationid = conversation_to_user.conversationid) WHERE conversation_to_user.participantid
= 1 AND conversation_to_user.hideconversation = 0 AND ( ( conversation.userid =
1 AND conversation.conversationid IN ( SELECT conversationid FROM
wcf2_conversation_to_user WHERE username IN ('test') GROUP BY conversationid
HAVING COUNT(conversationid) = 1 ) ) OR ( conversation.userid <> 1 AND
conversation.conversationid IN ( SELECT conversationid FROM
wcf2_conversation_to_user WHERE username IN ('test') AND isinvisible = 0 GROUP
BY conversationid HAVING COUNT(conversationid) = 1 ) ) ) 
```

```plain
+------+--------------+---------------------------+-------+-------------------------------------------------------------------+-------------------------------------+---------+------+------+-------------------------------------------------+
| id   | select_type  | table                     | type  | possible_keys                                                     | key                                 | key_len | ref  | rows | Extra                                           |
+------+--------------+---------------------------+-------+-------------------------------------------------------------------+-------------------------------------+---------+------+------+-------------------------------------------------+
|    1 | PRIMARY      | conversation              | index | PRIMARY,userID                                                    | userID                              | 6       | NULL |    4 | Using where; Using index                        |
|    1 | PRIMARY      | conversation_to_user      | ALL   | participantID,participantID_2,a6a806c449f2ed6688e5097fde0d145a_fk | NULL                                | NULL    | NULL |    9 | Using where; Using join buffer (flat, BNL join) |
|    3 | MATERIALIZED | wcf2_conversation_to_user | index | NULL                                                              | a6a806c449f2ed6688e5097fde0d145a_fk | 4       | NULL |    9 | Using where                                     |
|    2 | MATERIALIZED | wcf2_conversation_to_user | index | NULL                                                              | a6a806c449f2ed6688e5097fde0d145a_fk | 4       | NULL |    9 | Using where                                     |
+------+--------------+---------------------------+-------+-------------------------------------------------------------------+-------------------------------------+---------+------+------+-------------------------------------------------+
```

Reply:
I've also realized that there is currently no limit on the number of participants that can be included in the query, so that should be addressed too.